src/mlpack/methods/ann/rbm/spike_slab_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,344 @@
+
+  for (size_t i = 0; i < hiddenSize; i++)
+    spike(i) = math::RandBernoulli(spikeMean(i));
+}
+
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>::SlabMean(
+    DataType&& visible,
+    DataType&& spike,
+    DataType&& slabMean)
+{
+  assert(visible.n_rows == visibleSize);
+  assert(visible.n_cols == 1);
+
+  assert(spike.n_rows == hiddenSize);
+  assert(spike.n_cols == 1);
+
+  assert(slabMean.n_rows == poolSize);
+  assert(slabMean.n_cols == hiddenSize);
+
+  assert(weight.n_rows == visibleSize);
+  assert(weight.n_cols == poolSize);
+
+  for (size_t i = 0; i < hiddenSize; i++)
+  {
+    slabMean.col(i) = (1.0 / slabPenalty) * spike(i) *
+        weight.slice(i).t() * visible;
+  }
+}
+
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>::SampleSlab(
+    DataType&& slabMean,
+    DataType&& slab)
+{
+  assert(slabMean.n_rows == poolSize);
+  assert(slabMean.n_cols == hiddenSize);
+
+  assert(slab.n_rows == poolSize);
+  assert(slab.n_cols == hiddenSize);
+
+  for (size_t i = 0; i < hiddenSize; i++)
+  {
+    for (size_t j = 0; j < poolSize; j++)
+    {
+      slab(j, i) = math::RandNormal(slabMean(j, i), 1.0 / slabPenalty);
+    }
+  }
+}
+
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>::VisibleMean(
+    DataType&& input,
+    DataType&& output)
+{
+  assert(input.n_elem == hiddenSize + poolSize * hiddenSize);
+  output.set_size(visibleSize, 1);
+  output.zeros();
+
+  DataType spike(input.memptr(), hiddenSize, 1, false, false);
+  DataType slab(input.memptr() + hiddenSize, poolSize, hiddenSize, false,
+      false);
+
+  for (size_t i = 0; i < hiddenSize; i++)
+    output += weight.slice(i) * slab.col(i) * spike(i);
+
+  output = ((1.0 / visiblePenalty(0)) * output);
+}
+
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>::HiddenMean(
+    DataType&& input,
+    DataType&& output)
+{
+  assert(input.n_elem == visibleSize);
+  output.set_size(hiddenSize + poolSize * hiddenSize, 1);
+
+  DataType spike(output.memptr(), hiddenSize, 1, false, false);
+  DataType slab(output.memptr() + hiddenSize, poolSize, hiddenSize, false,
+      false);
+
+  SpikeMean(std::move(input), std::move(spike));
+  SampleSpike(std::move(spike), std::move(spikeSamples));
+  SlabMean(std::move(input), std::move(spikeSamples), std::move(slab));
+}
+
+template<typename DataType>
+void SpikeSlabRBMPolicy<DataType>::SampleVisible(
+    DataType&& input,
+    DataType&& output)
+{
+  const size_t numMaxTrials = 10;
+  size_t k = 0;
+
+  VisibleMean(std::move(input), std::move(visibleMean));
+
+  output.set_size(visibleSize, 1);
+
+  assert(visiblePenalty(0) > 0);
+
+  for (k = 0; k < numMaxTrials; k++)
+  {
+    for (size_t i = 0; i < visibleSize; i++)
+    {
+      output(i) = math::RandNormal(visibleMean(i), 1.0 / visiblePenalty(0));
+    }
+    if (arma::norm(output, 2) < radius)
+      break;
+  }
+
+  if (k == numMaxTrials)
+  {
+    Log::Warn << "Outputs are still not in visible unit"
```

Comment:
Missing space after ``unit`` and before ``terminating`` also no need to align the streams here (tab twice is fine).